editor/components/block-settings-menu/block-settings-menu-plugins-item.js

```diff
@@ -0,0 +1,39 @@
+/**
+ * External dependencies
+ */
+import { difference } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { IconButton } from '@wordpress/components';
+import { compose } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import BlockSettingsMenuPluginsGroup from './block-settings-menu-plugins-group';
+
+const shouldRenderItem = ( selectedBlockNames, allowedBlockNames ) => ! Array.isArray( allowedBlockNames ) ||
+	difference( selectedBlockNames, allowedBlockNames ).length === 0;
+
+const BlockSettingsMenuPluginsItem = ( { allowedBlocks, icon, label, onClick, small, role } ) => (
```

Comment:
Plugins may want to have multiple menu items and even have their own separators to separate the items. In that case, they would need to be repeating the allowedBlocks prop on each item plus use the BlockSettingsMenuPluginsGroup to pass the separator.
Maybe we can consider exposing the slot in a way that does not render anything and receives the allowed block props. And exposes a simple component called SettingMenuItem that just uses IconButton and handles the mobile label logic.
E.g:
```
<PluginBlockSettingsMenu allowedBlocks={ [...] }>
<BlockSettingMenuItem
	icon={}
	label="Item of the first group"
	onClick={}
/>
...
	<div className="editor-block-settings-menu__separator" />
<BlockSettingMenuItem
	icon={}
	label="Item of the second group"
	onClick={}
/>
...
</PluginBlockSettingsMenu>
```
This would also allow plugins more options to customize their UI.

Reply:
I like the idea of enabling plugins to introduce more than one item on the block settings menu.

How about this API?

```
<BlockSettingsMenuPluginsItem 
            allowedBlockNames=[ 'core/paragraph' ]
            secondaryMenu=[
                {
                  label: 'Secondary 1',
                  onClick: { secondary1OnClick }
                },
                {
                  label: 'Secondary 2',
                  onClick: { secondary2OnClick }
                },
            ],
            icon='dashicon-name'
            label='Plugin name'
            onClick={ primaryOnClick } />
```
The idea would be to enable plugins to add either one or multiple items, in an browsers-like fashion (compare `LastPass` VS `Save to Pocket` items):

![chrome-plugins-menu](https://user-images.githubusercontent.com/583546/42757657-1bd4b950-8901-11e8-98c5-27df50eeaf57.png)

If `BlockSettingsMenuPluginsItem` receives a `submenu` property it'll introduce a secondary menu and an arrow in the item to communicate there is another level. If `submenu` is not present it'll render a normal item. On mobile, we may want to use a different pattern (for ex: render the two items consecutively in the primary menu).

This doesn't address allowing plugins to add separators. I lean towards thinking separators should be UI only core can add - by staying the same no matter the plugins an user has installed Gutenberg communicates clearly how the menu is organized.

What do you think?

Reply:
Hi @nosolosw, I like the idea of submenus so plugins can group their actions if they have multiple operations.
This proposal introduces a new type of menus that we don't have in Gutenberg yet cc: @nosolosw.

Regarding the technical point of view:
What if we have a component that allows the rendering of multiple items like:
```
<MultipleItemsMenu
	icon='dashicon-name'
    label='Plugin name'"
>
	<MenuItem
		label="Secondary 1"
		onClick={}
	/>
	<MenuItem
		label="Secondary 2"
		onClick={}
	/>
</MultipleItemsMenu>
```


It would be a simple UI component that even core could use.

If we allow plugins to render what they want to the slot, then they could use this component as:
```
<PluginBlockSettingsMenu allowedBlocks={ [...] }>
<MultipleItemsMenu
	icon='dashicon-name'
    label='Plugin name'"
>
...
</PluginBlockSettingsMenu>
```

Both approaches are similar both if we use react components we have some things simplified.
Using `secondaryMenu=[ { label: 'Secondary 1', onClick: { secondary1OnClicl } ...,` Will create a new array reference and new object references per item inside the array. For plugins to avoid rerenders they will need to have some logic to cache the references and only update the references when onClick or labels change. If we represent the same info as React components that logic is not needed.

Reply:
As per private conversations with Tammie, she suggested that plugins shouldn't be able to have the flexibility to customize the slot at this point (adding separators, etc).

There are a couple of extra considerations we haven't considered yet and we should take into account:

* the API surface of this should be similar to the one of `PluginSidebarMenuItem`, as to create a coherent experience for developers.
* plugins may want to have a different set of `allowedBlocks` per item.

All things considered, I think we should go with the API as it is in the PR, as it meets all requirements (plugins can have more than one item, items can have different `allowedBlocks` sets, developers can't add their own separators/components, and the API shape is the same than the `PlugiSidebarMenuItem`). I still think a component for secondary menus could be useful, but that can be added later if it proves necessary/useful.

Thanks for your time, Jorge! This conversation has helped me to understand better Gutenberg APIs.